util/topgen/merge.py

```diff
@@ -296,6 +335,140 @@ def amend_xbar(top, xbar):
         xbar_adddevice(top, topxbar, device)
 
 
+def xbar_cross(xbar, xbars):
+    """Check if cyclic dependency among xbars
+
+    And gather the address range for device port (to another Xbar)
+
+    @param node_name if not "", the function only search downstream
+                     devices starting from the node_name
+    @param visited   The nodes it visited to reach this port. If any
+                     downstream port from node_name in visited, it means
+                     circular path exists. It should be fatal error.
+    """
+    # Step 1: Visit devices (gather the address range)
+    log.info("Processing circular path check for {}".format(xbar["name"]))
+    addr = []
+    for node in [
+            x for x in xbar["nodes"]
+            if x["type"] == "device" and "xbar" in x and x["xbar"] == False
```

Comment:
if the xbar happens to not be there for some reason, do we end up not gathering the specific device's address? Just wondering if it makes more sense in the validate step to make sure a xbar variable is always there and just assume its okay here.

Reply:
Yeah right. I've made a code to add "xbar" field for non xbar devices too. We could make it sure in `validate.py` code.